crates/bevy_ecs/src/world/entity_ref.rs

```diff
@@ -705,8 +687,36 @@ impl<'w> EntityWorldMut<'w> {
             return None;
         }
 
-        let mut bundle_components = bundle_info.components().iter().cloned();
         let entity = self.entity;
+        // SAFETY: Archetypes and Bundles cannot be mutably aliased through DeferredWorld
+        let (old_archetype, bundle_info, mut deferred_world) = unsafe {
+            let bundle_info: *const BundleInfo = bundle_info;
+            let world = world.as_unsafe_world_cell();
+            (
+                &world.archetypes()[old_location.archetype_id],
+                &*bundle_info,
+                world.into_deferred(),
+            )
+        };
+
+        if old_archetype.has_on_remove() {
+            // SAFETY: All components in the archetype exist in world
+            unsafe {
+                deferred_world.trigger_on_remove(entity, bundle_info.iter_components());
+            }
+        }
```

Comment:
It's not actually sound to keep `&Archetype` and `&BundleInfo` around while an `&mut World` exists. Even if you don't actually modify archetypes or bundles, the existence of the `&mut World` allows the compiler to break things.

This is my fault though -- this was correct before I suggested changing `DeferredWorld(UnsafeWorldCell)` to `DeferredWorld(&mut World)`.

Reply:
Oh yeah, that was why I had it that way in the first place. Can easily swap back.